src/core/CorePool.ts

```diff
@@ -157,11 +157,53 @@ export class CorePool {
     amount0Requested: JSBI,
     amount1Requested: JSBI
   ): { amount0: JSBI; amount1: JSBI } {
-    // TODO
     this.checkTicks(tickLower, tickUpper);
+
+    assert(
+      JSBI.greaterThanOrEqual(amount0Requested, ZERO),
+      "amount0Request should equal or greater than 0"
+    );
+    assert(
+      JSBI.greaterThanOrEqual(amount1Requested, ZERO),
+      "amount1Request should equal or greater than 0 "
+    );
+
+    let positionForCheck = this.positionManager.getPositionReadonly(
+      recipient,
+      tickLower,
+      tickUpper
+    );
+    assert(
+      JSBI.notEqual(positionForCheck.tokensOwed0, ZERO) ||
+        JSBI.notEqual(positionForCheck.tokensOwed1, ZERO),
+      "position is not collectable!"
+    );
+
+    assert(
+      JSBI.notEqual(positionForCheck.liquidity, ZERO),
+      "liquidity should NOT be 0"
+    );
```

Comment:
Actually it's necessary for position with 0 liquidity to be collectable. First, there will be 0 liquidity left after the caller burn all his/her liquidity on the position when the caller get no tokens, and tokensOwed stands for token amount one can get by calling `collect`. Second, one can still call `collect` on position with zero tokensOwed, though nothing happens. I guess there are such trxs in mainnet events. So in last comment it's uninitialized position to mis-trigger by mis-input from caller that I'm concerned about, we need to ensure position to collect from `collect` input has been initialized, otherwise we throw an error instead of initializing that(line 187). Since `PositionManager` doesn't offer a `isInitialized` check, I make the workaround as asserting `liquidity`, `tokensOwed0` and `tokensOwed1` aren't equal zero simultaneously, or the position should be uninitialized. Feel free to contact me if there is any issue left.

Reply:
https://github.com/Bella-DeFinTech/uniswap-v3-simulator/pull/35

`positionManager` now offers `isInitialized(key)` and `clear(key)`

Reply:
Cool! place all checks and value validation within positionManager is very clean. refine code based on pull #35

Reply:
https://github.com/Bella-DeFinTech/uniswap-v3-simulator/blob/c42f4f7b33ddd3e5506d947beaa1980e5025f61f/src/manager/PositionManager.ts#L74

Please note that `updateBurn` is handled already within `positionManager.collectPosition`

Reply:
Got it!